openedx/core/djangoapps/credentials_service/models.py

```diff
@@ -1,3 +1,273 @@
+
+
+def validate_image(image):
+    """
+    Validates that a particular image is small enough, of the right type and
+    square to be a badge.
+    """
+    if image.width != image.height:
+        raise ValidationError(_(u"The image must be square."))
+    if image.size > (250 * 1024):
+        raise ValidationError(_(u"The image file size must be less than 250KB."))
+
+
+class Signatory(TimeStampedModel):
+    """
+    Signatory model to add certificate signatories.
+    """
+    name = models.CharField(max_length=255)
+    title = models.CharField(max_length=255)
+    image = models.ImageField(
+        help_text=_(
+            u"Image must be square PNG files. The file size should be under 250KB."
+        ),
+        upload_to=assets_path,
+        validators=[validate_image]
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return u"{name}, {title}".format(
+            name=self.name,
+            title=self.title,
+        )
+
+    def save(self, *args, **kwargs):
+        """save the Signatory."""
+        if self.pk is None:
+            temp_image = self.image
+            self.image = None
+            super(Signatory, self).save(*args, **kwargs)
+            self.image = temp_image
+
+        super(Signatory, self).save(*args, **kwargs)
+
+
+class CertificateTemplate(TimeStampedModel):
+    """
+    Certificate Template model to organize content for certificates.
+    """
+
+    name = models.CharField(max_length=255, db_index=True)
+    content = models.TextField(
+        help_text=_(u"HTML Template content data.")
+    )
+    certificate_type = models.CharField(
+        max_length=255, null=True, blank=True,
+        choices=MODE_CHOICES,
+    )
+    organization_id = models.IntegerField(
+        null=True,
+        blank=True,
+        db_index=True,
+        help_text=_(u'Organization with which this template is attached.'),
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return u"{name}".format(
+            name=self.name
+        )
+
+
+class AbstractCertificate(AbstractCredential):
+    """
+    Abstract Certificate configuration to support multiple type of certificates
+    i.e. Programs, Courses.
+    """
+    signatory = models.ManyToManyField(Signatory)
+    template = models.ManyToManyField(CertificateTemplate, null=True, blank=True)
+    title = models.CharField(
+        max_length=255, null=True, blank=True,
+        help_text="Custom certificate title to override default display_name for a course/program."
+    )
+
+    class Meta(object):
+        abstract = True
+
+
+class CourseCertificate(AbstractCertificate):
+    """
+    Configuration for Course Certificates.
+    """
+    course_id = CourseKeyField(max_length=255)
+    certificate_type = models.CharField(
+        max_length=255,
+        choices=MODE_CHOICES
+    )
+
+    class Meta(object):
+        # Enforce the constraint that mode, certificate and site should be
```

Comment:
Don't write comments that simply describe the code. We all know Python/Django. http://google.github.io/styleguide/pyguide.html#Comments